### PR TITLE
Optimize FullyConnected FakeQuantize post-ops

### PR DESCRIPTION
**Pull request by mangguo321**
### Details:
 - *Optimize FullyConnected FakeQuantize post-ops with a inner product output scaling*
 - *Optimize FullConnected "eltwise+ FakeQuantize" post-ops with a eltwise scaling*

### Tickets:
 - *CVS-86027*
 - *CVS-85020*

### CODE REVIEW

**Comment by zhangYiIntel** [comment]
Hi Mang,
also need this to avoid bias fused with FQ
https://github.com/mangguo321/openvino/pull/8

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp [hunk 1]
```diff
@@ -393,9 +393,142 @@ void FullyConnected::setPostOps(dnnl::primitive_attr &attr, const VectorDims &di
...
+
+                        for (int i = 0; i < std::max(ch.size(), isc.size()); i++) {
+                            if (std::abs(ch[ch.size() == 1 ? 0 : i] * isc[isc.size() == 1 ? 0 : i] - 127.f) > 0.0001f) {
+                                isCropAligned = false;
+                            }
+                        }
+
+                        if (isCropAligned) {
+                            auto len = ops.len();
+                            ops.get()->entry_[len - 1].eltwise.scale = isc[0];
```
[the remark this answers is not in the document]
Comment (mangguo321) [reply src/plugins/intel_cpu/src/nodes/fullyconnected.cpp]: Done.

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp [hunk 2]
```diff
@@ -393,9 +393,142 @@ void FullyConnected::setPostOps(dnnl::primitive_attr &attr, const VectorDims &di
...
+                        }
+                    }
+
+                    if (outputDataType == memory::data_type::s8 &&
+                        std::all_of(ish.cbegin(), ish.cend(), [](float val) { return std::abs(val - 128.f) < 0.0001f; }) &&
+                        std::all_of(osc.cbegin(), osc.cend(), [](float val) { return val == 1.f; }) &&
+                        std::all_of(osh.cbegin(), osh.cend(), [](float val) { return std::abs(val + 128.f) < 0.0001f; })) {
+                        bool isCropAligned = true;
+                        for (int i = 0; i < std::max(cl.size(), isc.size()); i++) {
+                            if (std::abs(cl[0] * isc[0] + 128.f) > 0.0001f) {
```
[the remark this answers is not in the document]
Comment (mangguo321) [reply src/plugins/intel_cpu/src/nodes/fullyconnected.cpp]: Done.

File: src/plugins/intel_cpu/src/docs/fake_quantize.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+thus inputLow/inputHigh/outputLow/outputHigh is gently tuned from statistical result to satisfy these requirements.
+
+# Symetric quantization
+In symetric quantize: choose `il` to be `-ih` results in non-integer zero points (since levels is even number)
+$$
+   Z_i = \frac{-il*(levels-1)}{ih - il} = (levels-1)/2
+$$
+
+in symetric quantization, Zi is choosen to be `levels/2`, thus we can increase the range a little by push il to be smaller number
+$$
```
[the remark this answers is not in the document]
Comment (mangguo321) [reply src/plugins/intel_cpu/src/docs/fake_quantize.md:82]: Done.

File: src/plugins/intel_cpu/src/docs/fake_quantize.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+else:
+    # input_low < x <= input_high
+    output = round((x - input_low) / (input_high - input_low) \* (levels-1)) / (levels-1) \* (output_high - output_low) + output_low
+```
+
+ - x <= min(input_low, input_high): output = output_low
+ - x > max(input_low, input_high):  output = output_high
+ - input_low < x <= input_high:
+
+$$
```
Comment (xuchen-intel) [inline src/plugins/intel_cpu/src/docs/fake_quantize.md:19]:
```suggestion
$$
\begin{align}
   q = round(\frac{x - il}{ih - il} * (levels-1)) \\
   output = q * \frac{oh - ol}{levels-1}  + ol
\end{align}
$$
```

File: src/plugins/intel_cpu/src/docs/fake_quantize.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+   q = round(\frac{x - il}{ih - il} * (levels-1)) \\
+   output = q * \frac{oh - ol}{levels-1}  + ol
+$$
+
+simplified, suppose (ih > il):
+$$
```
Comment (xuchen-intel) [inline src/plugins/intel_cpu/src/docs/fake_quantize.md:25]:
```suggestion

$$
```

File: src/plugins/intel_cpu/src/docs/fake_quantize.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+   q' &= round(x*\frac{1}{S_i} + Z_i - Z_0)   \\
+   q_{I} &= clamp(q', -Z_0, Z_0-1) \\
+   output &= (q_{I} + Z_0 -  Z_{out})* S_o
+\end{align}
+$$
+
+here the center value Z0 is substracted before clamp to make a signed quantized value $q_I$ and it was added back later after clamp for mathematical equivalence.
+
+notice:
+ - equation (5) is traditional quantization x into q only if Zi is integer:
```
Comment (xuchen-intel) [inline src/plugins/intel_cpu/src/docs/fake_quantize.md:68]: Please mark the corresponding equation (5) and (7).

File: src/plugins/intel_cpu/src/docs/fake_quantize.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+# Symetric quantization
+In symetric quantize: choose `il` to be `-ih` results in non-integer zero points (since levels is even number)
+
+$$
+   Z_i = \frac{-il*(levels-1)}{ih - il} = (levels-1)/2
+$$
+
+in symetric quantization, Zi is choosen to be `levels/2`, thus we can increase the range a little by push il to be smaller number
+
+$$
```
Comment (xuchen-intel) [inline src/plugins/intel_cpu/src/docs/fake_quantize.md:82]:
```suggestion
$$
\begin{align}
   (levels-1)/Z_i = -(ih - il)/il = 1 - ih/il \\
   2(1-1/levels) = 1 - ih/il \\
   il = -ih/(1 - 2/levels)
\end{align}
$$
```

**Review by xuchen-intel — CHANGES_REQUESTED** [review]

**Review by usstq — APPROVED** [review]

**Review by xuchen-intel — APPROVED** [review]

**Pull request merged**